cmd/storage-class.go

```diff
@@ -0,0 +1,213 @@
+// If storage class is set using the env vars MINIO_STORAGE_CLASS_RRS and MINIO_STORAGE_CLASS_STANDARD
+// -- corresponding values are returned
+// If storage class is not set using environment variables, default values are returned
+// -- Default for Reduced Redundancy Storage class is, parity = 2 and data = N-Parity
+// -- Default for Standard Storage class is, parity = N/2, data = N/2
+// If storage class is not present in metadata, default value is data = N/2, parity = N/2
+func getDrivesCount(sc string, disks []StorageAPI) (data, parity int) {
+	totalDisks := len(disks)
+	switch sc {
+	case reducedRedundancyStorageClass:
+		if globalRRStorageClass.Parity != 0 {
+			// set the rrs parity if available
+			parity = globalRRStorageClass.Parity
+		} else {
+			// else fall back to default value
+			parity = defaultRRSParity
+		}
+	case standardStorageClass:
+		if globalStandardStorageClass.Parity != 0 {
+			// set the standard parity if available
+			parity = globalStandardStorageClass.Parity
+		} else {
+			// else fall back to default value
+			parity = totalDisks / 2
+		}
+	default:
+		parity = totalDisks / 2
+	}
+	// data is always totalDisks - parity
+	return totalDisks - parity, parity
+}
+
+// Returns per object readQuorum and writeQuorum
+// readQuorum is the minimum required disks to read data.
+// writeQuorum is the minimum required disks to write data.
+func objectQuorumFromMeta(xl xlObjects, partsMetaData []xlMetaV1, errs []error) (objectReadQuorum, objectWriteQuorum int, err error) {
+
+	// get the latest updated XL Metadata
+	latestXLMeta := listLatestXLMeta(partsMetaData, errs)
+
+	var validXLMeta xlMetaV1
+	emptyXLMetaV1 := xlMetaV1{}
+	count := 0
+
+	// latestXLMeta is a slice of xlMeta(s) updated most recently.
+	// We implicitly assume that all the xlMeta(s) have same dataBlocks and parityBlocks.
+	// Now loop through latestXLMeta and to get a count of valid xlMeta(s)
+	for _, xlMeta := range latestXLMeta {
+		if reflect.DeepEqual(xlMeta, emptyXLMetaV1) {
+			continue
+		} else {
+			count++
+			validXLMeta = xlMeta
+		}
+	}
+
+	// Simple majority of dataBlocks/parityBlocks from xl.json is not enough. We need to establish that
+	// at least dataBlocks number of xlMeta is available. This means count should be greater than or
+	// equal to dataBlocks field of validXLMeta. If not we through read quorum error.
+	if count < validXLMeta.Erasure.DataBlocks {
+		// This is the case when we can't reliably deduce object quorum
+		return 0, 0, errXLReadQuorum
+	}
+
+	// Since all the valid xl meta updated at the same time are equivalent, pass dataBlocks from one
+	// file to get the quorum
+	objectReadQuorum, objectWriteQuorum = objectQuorumFromDrives(validXLMeta.Erasure.DataBlocks)
+
+	return objectReadQuorum, objectWriteQuorum, nil
+}
+
+// returns readQuorum and writeQuorum based on number of blocks
+func objectQuorumFromDrives(dataBlocks int) (readQuorum, writeQuorum int) {
```

Comment:
Not sure if we need this function?

Reply:
Yes, let's collapse into function above.